addons/controllers/clusterbootstrap_controller.go

```diff
@@ -300,79 +343,371 @@ func (r *ClusterBootstrapReconciler) createOrPatchclusterBootstrapFromTemplate(
+	cbPkg *runtanzuv1alpha3.ClusterBootstrapPackage, clusterClient client.Client) (*kapppkgv1alpha1.Package, error) {
+
+	var err error
+	// Create or patch Package CR on remote cluster
+	localPackage := &kapppkgv1alpha1.Package{}
+	key := client.ObjectKey{Namespace: cluster.Namespace, Name: cbPkg.RefName}
+	if err = r.Client.Get(r.context, key, localPackage); err != nil {
+		// If there is an error to get the Carvel Package CR from local cluster, nothing needs to be created/cloned on remote.
+		// Let the reconciler try again.
+		r.Log.Error(err, fmt.Sprintf("unable to get package %s, nothing needs to be created on cluster %s/%s",
+			key.String(), cluster.Namespace, cluster.Name))
+		return nil, err
+	}
+	remotePackage := &kapppkgv1alpha1.Package{}
+	remotePackage.SetName(localPackage.Name)
+	// The Package CR on remote cluster needs to be under configured system namespace
+	remotePackage.SetNamespace(r.Config.SystemNamespace)
+	_, err = controllerutil.CreateOrPatch(r.context, clusterClient, remotePackage, func() error {
+		remotePackage.Spec = *localPackage.Spec.DeepCopy()
+		// TODO: Follow up to see if we need to preserve all the other fields, like annotations
+		// https://github.com/vmware-tanzu/tanzu-framework/issues/1678
+		return nil
+	})
+	if err != nil {
+		r.Log.Error(err, fmt.Sprintf("unable to create or patch Package resource %s/%s on cluster: %s/%s",
+			remotePackage.Namespace, remotePackage.Name, cluster.Namespace, cluster.Name))
+		return nil, err
+	}
+	return remotePackage, nil
+}
+
+// createOrPatchPackageInstallOnRemote creates or patches PackageInstall CR on remote cluster. The kapp-controller
+// running on remote cluster will reconcile it and deploy resources.
+func (r *ClusterBootstrapReconciler) createOrPatchPackageInstallOnRemote(cluster *clusterapiv1beta1.Cluster,
+	cbPkg *runtanzuv1alpha3.ClusterBootstrapPackage, remoteSecret *corev1.Secret, clusterClient client.Client) (*kapppkgiv1alpha1.PackageInstall, error) {
+
+	// In order to create PackageInstall CR, we need to get the Package.Spec.RefName and Package.Spec.Version
+	remotePackageRefName, remotePackageVersion, err := util.GetPackageMetadata(r.context, clusterClient, cbPkg.RefName, r.Config.SystemNamespace)
+	if remotePackageRefName == "" || remotePackageVersion == "" || err != nil {
+		// Package.Spec.RefName and Package.Spec.Version are required fields for Package CR. We do not expect them to be
+		// empty and error should not happen when fetching them from a Package CR.
+		r.Log.Error(err, fmt.Sprintf("unable to fetch Package.Spec.RefName or Package.Spec.Version from Package %s/%s on cluster %s/%s",
+			r.Config.SystemNamespace, cbPkg.RefName, cluster.Namespace, cluster.Name))
+		return nil, err
+	}
+
+	// Create PackageInstall CRs on the remote workload cluster, kapp-controller will take care of reconciling them
+	remotePkgi := &kapppkgiv1alpha1.PackageInstall{
+		ObjectMeta: metav1.ObjectMeta{
```

Comment:
Same as above for kapp pkgi, I believe we should also add a label containing the cluster-name. I can add that later